tutorials/mnist_tutorial_pure_tf.py

```diff
@@ -31,32 +32,56 @@
 """
 
 
-class MLP(object):
+class MLP(Model):
   """
   An example of a bare bones multilayer perceptron (MLP) class.
   """
 
   def __init__(self, layers, input_shape):
+    super(MLP, self).__init__()
+
+    self.layer_names = []
     self.layers = layers
     self.input_shape = input_shape
-    for layer in self.layers:
+    for i,layer in enumerate(self.layers):
+      name = layer.__class__.__name__
+      self.layer_names.append(name+"_"+str(i))
+
       layer.set_input_shape(input_shape)
       input_shape = layer.get_output_shape()
 
-  def fprop(self, x, return_all=False, set_ref=False):
+  def get_layer_names(self):
+    return self.layer_names
+
+  def fprop(self, x, set_ref=False):
+    if (x, self.state) in self.fprop_cache.keys():
+      return self.fprop_cache[x, self.state]
     states = []
     for layer in self.layers:
       if set_ref:
         layer.ref = x
       x = layer.fprop(x)
       assert x is not None
       states.append(x)
-    if return_all:
-      return states
-    return x
+    self.fprop_cache[x, self.state] = states
+    return states
```

Comment:
I thought the Model class API specified that fprop returns a dictionary. This is returning a list.